inspirehep/records/api/base.py

```diff
@@ -981,3 +1015,83 @@ def dumps_for_es(self):
     def get_value(self, field, default=None):
         """Method which makes ``get_value`` more intuitive"""
         return get_value(self, field, default)
+
+    def is_superseded(self):
+        """Checks if record is superseded
+
+        Returns:
+            bool: True if is superseded, False otherwise
+        """
+        if "successor" in self.get_value("related_records.relation", ""):
+            return True
+        return False
+
+    def _update_refs_in_citation_table(self, save_every=100):
+        """Updates all references in citation table.
+            First removes all references (where citer is this record),
+            then adds all from the record again.
+
+        Args:
+            save_every (int): How often data should be saved into session.
+                One by one is very inefficient, but also 10000 at once.
+        """
+        with db.session.begin_nested():
+            CitationTable.query.filter_by(CiterId=self.id).delete()
+
+        if (
+            self.is_superseded()
+            or self.get("deleted")
+            or self.pid_type not in ["lit", "dat"]
+        ):
+            # Record is not eligible to cite
+            return
+        records_pids = self.get_linked_pids_from_field("references.record")
+        records_data = self.get_records_ids_by_pids(records_pids)
+        referenced_records = []
+        references_waiting_for_commit = []
+        citation_date = self.get_earliest_date()
+        for reference in records_data:
+            if reference not in referenced_records:
+                referenced_records.append(reference)
+                references_waiting_for_commit.append(
+                    CitationTable(
+                        CiterId=self.model.id,
+                        CitedId=reference,
+                        CitationDate=citation_date,
+                    )
+                )
+            if len(references_waiting_for_commit) >= save_every:
+                with db.session.begin_nested():
+                    db.session.bulk_save_objects(references_waiting_for_commit)
+                    references_waiting_for_commit = []
+        if references_waiting_for_commit:
+            with db.session.begin_nested():
+                db.session.bulk_save_objects(references_waiting_for_commit)
+
+    def get_earliest_date(self):
+        """Returns earliest date. If earliest date is missing month or day
+            it's set as 1 as DB do not accept date without day or month"""
+        date_paths = [
+            "preprint_date",
+            "thesis_info.date",
+            "thesis_info.defense_date",
+            "publication_info.year",
+            "legacy_creation_date",
+            "imprints.date",
+        ]
+
+        dates = [
+            str(el)
+            for el in chain.from_iterable(
+                force_list(self.get_value(path)) for path in date_paths
+            )
+        ]
+        if dates:
+            date = earliest_date(dates)
```

Comment:
`earliest_date(dates)` is basically `min(PartialDate.loads(d) for d in dates).dumps()`, so no need to use `earliest_date` if you need to access the `PartialDate` later anyway. Besides, maybe it would be better to add a method to `PartialDate` to do the following logic instead of doing it here?

Reply:
Done https://github.com/inspirehep/inspire-utils/pull/53 :+1: